### PR TITLE
[blog] Blog post on Thanos

### PR DESCRIPTION
**Pull request by holger-wg2**

### CODE REVIEW

**Comment by jonny-wg2** [comment]
Great blog post @holger-wg2 .

Maybe if we added a section or further elaborated on how this benefits us at Working Group Two more. For example, with Thanos we are able to:
- analyze trends and identify anomalies within the mobile core network
- have access to desensitized data across all of our services
- visualize and graph service usage
- etc....

Maybe we can also include a screenshot of the architecture or grafana dashboards?

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+---
+layout: blogpost
+permalink: /blog/metrics-unlimited-thanos/
+title: "Towards Observability nirvana: Infinite metric retention with Thanos"
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Either more capitals or less (I prefer less) for the title :)

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+date: 2020-01-28
+tags: infrastructure observability prometheus thanos kubernetes
+author: <a href="https://www.linkedin.com/in/hihrig/">Holger Ihrig</a>
+---
+In nowadays DevOps world our industry is relying on the ability to observe and monitorize our infrastructure and
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: In the current DevOps world, our industry relies on

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
+---
+layout: blogpost
+permalink: /blog/metrics-unlimited-thanos/
+title: "Towards Observability nirvana: Infinite metric retention with Thanos"
+date: 2020-01-28
+tags: infrastructure observability prometheus thanos kubernetes
+author: <a href="https://www.linkedin.com/in/hihrig/">Holger Ihrig</a>
+---
+In nowadays DevOps world our industry is relying on the ability to observe and monitorize our infrastructure and 
+services quite a bit. Working Group Two is no exception at this and as we are operating in the TelCo space
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~quite a bit~

Comment (annaken) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Working Group Two is no exception here, and

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+we wanted to know more about the usage patterns of our platform throughout days, months and even years.
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~throughout~ over

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+Internally we have been running Prometheus for a long time with a fairly limited retention of 30 days. This did not
+allow us to look back far enough to make the observations we wanted to.
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: to look far enough back in time

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+Luckily for us there already was a solution out there that would fill our needs and in addition to that make our
+monitoring stack more resilient. The solution is called [Thanos](https://thanos.io/).
+
+## Thanos
+Thanos was originally developed by a company called [Improbable](https://improbable.io/) to achieve long term storage
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~achieve~ provide

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+for Prometheus. It evolved into a much more complicated component wildly improving the scalability of the
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: which wildly improved

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+Prometheus monitoring Stack.
+
+The basic functionality however is that Thanos will upload the metrics collected by Prometheus onto any service with a 
+S3-compatible API or any other Storage Target supported by the Prometheus Remote Write feature. For readability we
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: storage target and remote write don't need caps

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+will only refer to it as S3 Storage as this is our storage target.
+
+We shall briefly look at all those components before describing how we are leveraging Thanos to obtain a higher metric 
+retention and higher reliability.
+
+### Thanos Sidecar
+The Sidecar runs as the name suggests in the same Pod as Prometheus and observes when Prometheus stores new storage
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: same ~P~pod as Prometheus, and

Comment (annaken) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: observes when Prometheus creates (?) new storage

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+buckets on disk, which it does about every 2 hours. If configured to do so, it will upload those storage bucket into S3.
+Another important feature however is that it extends the Prometheus Pod with an API that can be used by Thanos Querier
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~however~

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+as a Store API endpoint to query Prometheus metrics.
+
+### Thanos Store
+Thanos Store implements the Thanos Store API and makes the metric data stored in the S3 bucket available to the
+Thanos Querier. To do that it observes the configured S3 Bucket and reads the Metadata of the stored storage buckets
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~M~metadata

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+available in S3.
+
+### Thanos Querier
+Querier implements the Prometheus Query API and understands PromQL. It then sends the query using the aforementioned
+Store API to all known Thanos Stores (discovered using service discovery) and awaits the metric information from the
+stores, be it directly from Prometheus via the Sidecar or metrics stored in S3 Object storage via Thanos Store.
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~S~sidecar

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+### Thanos Compactor
+It does not make a lot of sense to keep old metrics that are scraped every 15 or 30 seconds forever. At one point those
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~one~ some point ~those~these

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+according to the Recording Rules fed into the Ruler component. However since this rule processing is not done against a 
+local datastore, it is possible that these new metric datapoints will not always be generated as it relies on a reliable
+data source to do this in the required intervals.
+
+## Architecture in a Cluster
+
+![Thanos Architecture](/img/blog/thanos/thanos_architecture.png)
+
+As we can see, there are quite a few things going on in this architectural view of the system, but on the other hand it
+is fairly simple to understand as the component are nicely decoupled from each other.
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: components

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+An interesting thing here is that the drawing has multiple Prometheus instances with multiple Sidecars. Thanos actually
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~S~sidecars

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+allows for deduplication of Timeseries Data. The data uploaded from the sidecars have information from which prometheus
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: timeseries data

Comment (annaken) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: ~have information from which prometheus~ 
contains information about which prometheus

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+instance the metrics are generated and adds that. The Querier can then deduplicate this data so that the metrics shown
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: are generated from, and adds that

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+in Grafana are consistent and do not come sometimes from Prometheus A and sometimes from Prometheus B.
+
+This design also allows for an interesting other Use Case: Querying multiple Clusters. As
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: use case: querying multiple clusters (caps!)

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+long as the storage location used by the Sidecars for uploading the metrics is identical, the Time Series data is
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: sidecars

Comment (annaken) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: time series

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+- Highly Available Prometheus
+- Increased Reliability (Decoupled query component)
+- Infinite Metric Storage
+- Query Multiple clusters from a single point
+- Easy to Scale
+
+### Cons
+- More complex architecture
+- Increased Resource usage
+
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: So many caps in pros and cons :) 
I'd just capitalise Prometheus and nothing else

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+## Summary
+
+Even though Thanos comes with an increased architectural and operational complexity, we have to say after running
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: after running it for a while, we think it is totally worth it

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+with it for a while that it is totally worth it. We can make architecture decisions by looking further
+back in time than before. It also has the additional Advantage that a misconfiguration of a Prometheus Deployment
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: advantage
deployment

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+does not pull down the whole stack as the Prometheus is HA and only updates one at a time. If the deployment fails,
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: does not pull down the whole stack as ~the~ Prometheus is HA, and only one instance updates at a time.

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+- Increased Resource usage
+
+## Summary
+
+Even though Thanos comes with an increased architectural and operational complexity, we have to say after running
+with it for a while that it is totally worth it. We can make architecture decisions by looking further
+back in time than before. It also has the additional Advantage that a misconfiguration of a Prometheus Deployment
+does not pull down the whole stack as the Prometheus is HA and only updates one at a time. If the deployment fails, 
+we can get notified and the deployment can be aborted. Also upgrades of the whole monitoring stack, can
+now be done more gradual, which is also a great advantage for us!
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: gradually

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 5 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 4 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 3 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 2 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 1 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 6 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 7 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 8 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 9 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 10 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 11 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 12 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 13 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 14 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 15 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 16 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 17 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 18 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 19 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 20 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 21 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 22 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 23 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md (hunk 24 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: Done

File: pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md [hunk 25]
```diff
@@ -87,6 +87,17 @@ That even allows to run Grafana, Querier and Store in a completely different par
 - more complex architecture
 - increased resource usage
 
+## What are the biggest benefits for us?
+
+As stated in the beginning of this article one of our objectives for implementing Thanos was the increased
+metric retention to be able to look back further in time.
+
+These are the most interesting applications for us:
```
Comment (annaken) [inline pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md]: The most interesting features for us are to be able to:

**Pull request merged**